File: docs/core/examples/ruby/multiparty_trades.rb
require 'chain'

# This demo is written to run on either one or two cores. Simply provide
# different URLs to the following clients for the two-core version.
Client aliceCore = Chain::Client.new
Client bobCore = Chain::Client.new
signer = Chain::HSMSigner.new

alice_dollar_key = MockHsm.Key.create(aliceCore)
signer.add_key(alice_dollar_key, MockHsm.getSignerClient(aliceCore))

bob_buck_key = MockHsm.Key.create(bobCore)
signer.add_key(bob_buck_key, MockHsm.getSignerClient(bobCore))

alice_key = MockHsm.Key.create(aliceCore)
signer.add_key(alice_key, MockHsm.getSignerClient(aliceCore))

bob_key = MockHsm.Key.create(bobCore)
signer.add_key(bob_key, MockHsm.getSignerClient(bobCore))

Asset aliceDollar = chain.assets.create(
  alias: 'aliceDollar',
  root_xpubs: [alice_dollar_key.xpub],
  quorum: 1,
  .create(aliceCore)

Asset bobBuck = chain.assets.create(
  alias: 'bobBuck',
  root_xpubs: [bob_buck_key.xpub],
  quorum: 1,
  .create(bobCore)

Account alice = chain.accounts.create(
  alias: 'alice',
  root_xpubs: [alice_key.xpub],
  quorum: 1,
  .create(aliceCore)

Account bob = chain.accounts.create(
  alias: 'bob',
  root_xpubs: [bob_key.xpub],
  quorum: 1,
  .create(bobCore)

chain.transactions.submit(aliceCore, signer.sign(chain.transactions.build do |b|
  b.issue
    asset_alias: 'aliceDollar',
    amount: 1000,
  b.control_with_account
    account_alias: 'alice',
    asset_alias: 'aliceDollar',
    amount: 1000,
  ).build(aliceCore)
))

chain.transactions.submit(bobCore, signer.sign(chain.transactions.build do |b|
  b.issue
    asset_alias: 'bobBuck',
    amount: 1000,
  b.control_with_account
    account_alias: 'bob',
    asset_alias: 'bobBuck',
    amount: 1000,
  ).build(bobCore)
))

if (aliceCore.equals(bobCore)) {
  sameCore(aliceCore)
}

crossCore(aliceCore, bobCore, alice, bob, aliceDollar.id, bobBuck.id)
}

public static void sameCore(chain) throws Exception {
# snippet same-core-trade
trade = chain.transactions.build do |b|
  b.spend_from_account
    account_alias: 'alice',
    asset_alias: 'aliceDollar',
    amount: 50,
  b.control_with_account
    account_alias: 'alice',
    asset_alias: 'bobBuck',
    amount: 100,
  b.spend_from_account
    account_alias: 'bob',
    asset_alias: 'bobBuck',
    amount: 100,
  b.control_with_account
    account_alias: 'bob',
    asset_alias: 'aliceDollar',
    amount: 50,
  ).build(client)

chain.transactions.submit(signer.sign(trade))
# endsnippet
}

public static void crossCore(
Client aliceCore, Client bobCore,
Account alice, Account bob,
String aliceDollarAssetId, String bobBuckAssetId
) throws Exception {
# snippet build-trade-alice
aliceTrade = chain.transactions.build do |b|
  b.spend_from_account
    account_alias: 'alice',
    asset_alias: 'aliceDollar',
    amount: 50,
  b.control_with_account
    account_alias: 'alice',
    .setAssetId(bobBuckAssetId)
    amount: 100,
  ).build(aliceCore)
# endsnippet

# snippet sign-trade-alice
aliceTradeSigned = signer.sign(aliceTrade.allowAdditionalActions())
# endsnippet

# snippet base-transaction-alice
String baseTransactionFromAlice = aliceTradeSigned.rawTransaction
# endsnippet

# snippet build-trade-bob
bobTrade = chain.transactions.build do |b|
  .setBaseTransaction(baseTransactionFromAlice)
  b.spend_from_account
    account_alias: 'bob',
    asset_alias: 'bobBuck',
    amount: 100,
  b.control_with_account
    account_alias: 'bob',
    .setAssetId(aliceDollarAssetId)
    amount: 50,
  ).build(bobCore)
# endsnippet

# snippet sign-trade-bob
bobTradeSigned = signer.sign(bobTrade)
# endsnippet

# snippet submit-trade-bob
chain.transactions.submit(bobCore, bobTradeSigned)
# endsnippet

File: docs/core/examples/node/multipartyTrades.js
const chain = require('chain-sdk')

// This demo is written to run on either one or two cores. Simply provide
// different URLs to the following clients for the two-core version.
const aliceCore = new chain.Client()
const bobCore = new chain.Client()

const aliceSigner = new chain.HsmSigner()
const bobSigner = new chain.HsmSigner()

let aliceDollarKey, bobBuckKey, aliceKey, bobKey, aliceDollar, bobBuck

Promise.all([
  aliceCore.mockHsm.keys.create(),
  bobCore.mockHsm.keys.create(),
  aliceCore.mockHsm.keys.create(),
  bobCore.mockHsm.keys.create(),
]).then(keys => {
  aliceDollarKey  = keys[0].xpub
  bobBuckKey      = keys[1].xpub
  aliceKey        = keys[2].xpub
  bobKey          = keys[3].xpub

  aliceSigner.addKey(aliceDollarKey, aliceCore.mockHsm.signerConnection)
  bobSigner.addKey(bobBuckKey, bobCore.mockHsm.signerConnection)
  aliceSigner.addKey(aliceKey, aliceCore.mockHsm.signerConnection)
  bobSigner.addKey(bobKey, bobCore.mockHsm.signerConnection)
}).then(() => Promise.all([

  // snippet create-asset-aliceDollar
  aliceCore.assets.create({
    alias: 'aliceDollar',
    rootXpubs: [aliceDollarKey],
    quorum: 1,
  }),
  // endsnippet

  // create-asset-bobBuck
  bobCore.assets.create({
    alias: 'bobBuck',
    rootXpubs: [bobBuckKey],
    quorum: 1,
  }),
  // endsnippet

  // snippet create-account-alice
  aliceCore.accounts.create({
    alias: 'alice',
    rootXpubs: [aliceKey],
    quorum: 1,
  }),
  // endsnippet

  // snippet create-account-bob
  bobCore.accounts.create({
    alias: 'bob',
    rootXpubs: [bobKey],
    quorum: 1,
  })
  // endsnippet

])).then(assets => {
  aliceDollar = assets[0]
  bobBuck = assets[1]
}).then(() =>
  aliceCore.transactions.build(builder => {
    builder.issue({ assetAlias: 'aliceDollar', amount: 1000 })
    builder.controlWithAccount({
      accountAlias: 'alice',
      assetAlias: 'aliceDollar',
      amount: 1000
    })
  }).then(issuance => aliceSigner.sign(issuance))
    .then(signed => aliceCore.transactions.submit(signed))
).then(() =>
  bobCore.transactions.build(builder => {
    builder.issue({ assetAlias: 'bobBuck', amount: 1000 })
    builder.controlWithAccount({
      accountAlias: 'bob',
      assetAlias: 'bobBuck',
      amount: 1000
    })
  }).then(issuance => bobSigner.sign(issuance))
    .then(signed => bobCore.transactions.submit(signed))
).then(() => {
  if (aliceCore.baseUrl == bobCore.baseUrl){
    const chain = aliceCore
    const signer = aliceSigner
    signer.addKey(bobKey, chain.mockHsm.signerConnection)

    // SAME-CORE TRADE

    // snippet same-core-trade
    chain.transactions.build(builder => {
      builder.spendFromAccount({
        accountAlias: 'alice',
        assetAlias: 'aliceDollar',
        amount: 50
      })
      builder.controlWithAccount({
        accountAlias: 'alice',
        assetAlias: 'bobBuck',
        amount: 100
      })
      builder.spendFromAccount({
        accountAlias: 'bob',
        assetAlias: 'bobBuck',
        amount: 100
      })
      builder.controlWithAccount({
        accountAlias: 'bob',
        assetAlias: 'aliceDollar',
        amount: 50
      })
    }).then(trade => signer.sign(trade))
      .then(signed => chain.transactions.submit(signed))
    // endsnippet

  } else {
    // CROSS-CORE TRADE

    const aliceDollarAssetId = aliceDollar.id
    const bobBuckAssetId = bobBuck.id

    // snippet build-trade-alice
    aliceCore.transactions.build(builder => {
      builder.spendFromAccount({
        accountAlias: 'alice',
        assetAlias: 'aliceDollar',
        amount: 50
      })
      builder.controlWithAccount({
        accountAlias: 'alice',
        assetId: bobBuckAssetId,
        amount: 100
      })
    })
    // endsnippet

      // snippet sign-trade-alice
      .then(aliceTrade => {
        aliceTrade.allowAdditionalActions = true
        return aliceSigner.sign(aliceTrade)
      })
      // endsnippet

      .then(aliceSigned =>

        // snippet build-trade-bob
        bobCore.transactions.build(builder => {
          builder.baseTransaction(aliceSigned.rawTransaction)
          builder.spendFromAccount({
            accountAlias: 'bob',
            assetAlias: 'bobBuck',
            amount: 100
          })
          builder.controlWithAccount({
            accountAlias: 'bob',
            assetId: aliceDollarAssetId,
            amount: 50
          })
        })
        // endsnippet

        // snippet sign-trade-bob
        .then(bobTrade => bobSigner.sign(bobTrade))
        // endsnippet

        // snippet submit-trade-bob
        .then(bobSigned => bobCore.transactions.submit(bobSigned))
        // endsnippet
    )
  }
}).catch(err =>
  process.nextTick(() => { throw err })
)
